src/solvers/strings/string_constraint_generator_format.cpp

```diff
@@ -365,6 +366,55 @@ add_axioms_for_format_specifier(
   }
 }
 
+/// Deserialize an argument for format from \p string.
+/// \p id should be one of: string_expr, int, char, boolean, float.
+/// The primitive values are expected to all be encoded using 4 characters.
+static exprt format_arg_from_string(
+  const array_string_exprt &string, const irep_idt &id)
+{
+  if(id == "string_expr")
+    return string;
+  if(id == ID_int)
+  {
+    // Assume the string has length 4
+    // (int64)string.content[0] << 48 | (int64) string.content[1] << 32 |
+    // (int64)string.content[2] << 16 | (int64) string.content[3]
+    const signedbv_typet type{64};
+    return bitor_exprt{
+      bitor_exprt{
+        shl_exprt{typecast_exprt{string[0], type}, 48},
+        shl_exprt{typecast_exprt{string[1], type}, 32}},
+      bitor_exprt{
+        shl_exprt{typecast_exprt{string[2], type}, 16},
+        typecast_exprt{string[3], type}}};
+  }
+  if(id == ID_char)
+  {
+    // We assume the string has length exactly 4 and ignore the first 3
+    // (unsigned16)string.content[3]
+    const unsignedbv_typet type{16};
+    return typecast_exprt{
+      index_exprt{string.content(), from_integer(3, string.length().type())},
+      type};
+  }
+  if(id == ID_boolean)
+  {
+    // We assume the string has length exactly 4 and ignore the first 3
+    // (bool)string.content[3]
+    const c_bool_typet type{8};
+    return typecast_exprt{
+      index_exprt{string.content(), from_integer(3, string.length().type())},
+      type};
+  }
+  if(id == ID_float)
+  {
+    // Deserialize an int and cast to float
+    const exprt as_int = format_arg_from_string(string, ID_int);
+    return typecast_exprt{as_int, floatbv_typet{}};
+  }
+  UNHANDLED_CASE;
```

Comment:
byte, short, double and long are not handled yet?

Reply:
They are all handled by being cast to a long within the model. So ID_int does not need to make a distinction between byte, short, int and long.

Reply:
Drive-by comment: so what about `double`?

Reply:
It's also cast to a long. ID_float handles both the cases where the formatting string except a float and a double (this was the same before my PR).